// @wordpress/components/build/validated-form-controls/components/form-token-field.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ValidatedFormTokenField = void 0;
var _element = require("@wordpress/element");
var _controlWithError = require("../control-with-error");
var _formTokenField = require("../../form-token-field");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const UnforwardedValidatedFormTokenField = ({
  required,
  onValidate,
  customValidity,
  onChange,
  markWhenOptional,
  ...restProps
}, forwardedRef) => {
  const validityTargetRef = (0, _element.useRef)(null);
  const valueRef = (0, _element.useRef)(restProps.value);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    className: "components-validated-control__wrapper-with-error-delegate",
    ref: forwardedRef,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_controlWithError.ControlWithError, {
      required: required,
      markWhenOptional: markWhenOptional,
      onValidate: () => {
        return onValidate?.(valueRef.current);
      },
      customValidity: customValidity,
      getValidityTarget: () => validityTargetRef.current,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_formTokenField.FormTokenField, {
        __next40pxDefaultSize: true,
        __nextHasNoMarginBottom: true,
        ...restProps,
        onChange: (value, ...args) => {
          valueRef.current = value;
          onChange?.(value, ...args);
        }
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("input", {
      className: "components-validated-control__error-delegate",
      type: "text",
      ref: validityTargetRef,
      required: required,
      value: valueRef.current && valueRef.current.length > 0 ? 'hasvalue' : '',
      tabIndex: -1,
      onChange: () => {},
      onFocus: e => {
        e.target.previousElementSibling?.querySelector('input[type="text"]')?.focus();
      }
    })]
  });
};
const ValidatedFormTokenField = exports.ValidatedFormTokenField = (0, _element.forwardRef)(UnforwardedValidatedFormTokenField);
