// @wordpress/components/build/validated-form-controls/components/toggle-group-control.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ValidatedToggleGroupControl = void 0;
var _element = require("@wordpress/element");
var _controlWithError = require("../control-with-error");
var _toggleGroupControl = require("../../toggle-group-control");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const UnforwardedValidatedToggleGroupControl = ({
  required,
  onValidate,
  customValidity,
  onChange,
  markWhenOptional,
  ...restProps
}, forwardedRef) => {
  const validityTargetRef = (0, _element.useRef)(null);
  const valueRef = (0, _element.useRef)(restProps.value);
  const nameAttr = (0, _element.useId)();
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    className: "components-validated-control__wrapper-with-error-delegate",
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_controlWithError.ControlWithError, {
      required: required,
      markWhenOptional: markWhenOptional,
      onValidate: () => {
        return onValidate?.(valueRef.current);
      },
      customValidity: customValidity,
      getValidityTarget: () => validityTargetRef.current,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_toggleGroupControl.ToggleGroupControl, {
        __nextHasNoMarginBottom: true,
        __next40pxDefaultSize: true,
        ref: forwardedRef,
        onChange: value => {
          valueRef.current = value;
          onChange?.(value);
        },
        ...restProps
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("input", {
      className: "components-validated-control__error-delegate",
      type: "radio",
      ref: validityTargetRef,
      required: required,
      checked: restProps.value !== undefined,
      tabIndex: -1
      // A name attribute is needed for the `required` behavior to work.
      ,
      name: nameAttr,
      onChange: () => {},
      onFocus: e => {
        e.target.previousElementSibling?.querySelector('[data-active-item="true"]')?.focus();
      }
    })]
  });
};
const ValidatedToggleGroupControl = exports.ValidatedToggleGroupControl = (0, _element.forwardRef)(UnforwardedValidatedToggleGroupControl);
