// @wordpress/components/build/validated-form-controls/validity-indicator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ValidityIndicator = ValidityIndicator;
var _clsx = _interopRequireDefault(require("clsx"));
var _icons = require("@wordpress/icons");
var _icon = _interopRequireDefault(require("../icon"));
var _spinner = _interopRequireDefault(require("../spinner"));
var _jsxRuntime = require("react/jsx-runtime");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function ValidityIndicator({
  type,
  message
}) {
  const ICON = {
    valid: _icons.published,
    invalid: _icons.error
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("p", {
    className: (0, _clsx.default)('components-validated-control__indicator', `is-${type}`),
    children: [type === 'validating' ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_spinner.default, {
      className: "components-validated-control__indicator-spinner"
    }) : /*#__PURE__*/(0, _jsxRuntime.jsx)(_icon.default, {
      className: "components-validated-control__indicator-icon",
      icon: ICON[type],
      size: 16,
      fill: "currentColor"
    }), message]
  });
}
