// @wordpress/block-editor/build/components/spacing-sizes-control/linked-button.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LinkedButton;
var _components = require("@wordpress/components");
var _icons = require("@wordpress/icons");
var _i18n = require("@wordpress/i18n");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * WordPress dependencies
 */

function LinkedButton({
  isLinked,
  ...props
}) {
  const label = isLinked ? (0, _i18n.sprintf)(
  // translators: 1. Type of spacing being modified (padding, margin, etc).
  (0, _i18n.__)('Unlink %1$s'), props.label.toLowerCase()).trim() : (0, _i18n.sprintf)(
  // translators: 1. Type of spacing being modified (padding, margin, etc).
  (0, _i18n.__)('Link %1$s'), props.label.toLowerCase()).trim();
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Tooltip, {
    text: label,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_components.Button, {
      ...props,
      size: "small",
      icon: isLinked ? _icons.link : _icons.linkOff,
      iconSize: 24,
      "aria-label": label
    })
  });
}
