// @wordpress/components/build-module/dropdown-menu-v2/radio-item.js
/**
 * External dependencies
 */
import * as Ariakit from '@ariakit/react';

/**
 * WordPress dependencies
 */
import { forwardRef, useContext } from '@wordpress/element';
import { Icon } from '@wordpress/icons';

/**
 * Internal dependencies
 */

import { DropdownMenuContext } from './context';
import * as Styled from './styles';
import { SVG, Circle } from '@wordpress/primitives';
import { useTemporaryFocusVisibleFix } from './use-temporary-focus-visible-fix';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
const radioCheck = /*#__PURE__*/_jsx(SVG, {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 24 24",
  children: /*#__PURE__*/_jsx(Circle, {
    cx: 12,
    cy: 12,
    r: 3
  })
});
export const DropdownMenuRadioItem = forwardRef(function DropdownMenuRadioItem({
  suffix,
  children,
  onBlur,
  hideOnClick = false,
  ...props
}, ref) {
  // TODO: Remove when https://github.com/ariakit/ariakit/issues/4083 is fixed
  const focusVisibleFixProps = useTemporaryFocusVisibleFix({
    onBlur
  });
  const dropdownMenuContext = useContext(DropdownMenuContext);
  return /*#__PURE__*/_jsxs(Styled.DropdownMenuRadioItem, {
    ref: ref,
    ...props,
    ...focusVisibleFixProps,
    accessibleWhenDisabled: true,
    hideOnClick: hideOnClick,
    store: dropdownMenuContext?.store,
    children: [/*#__PURE__*/_jsx(Ariakit.MenuItemCheck, {
      store: dropdownMenuContext?.store,
      render: /*#__PURE__*/_jsx(Styled.ItemPrefixWrapper, {})
      // Override some ariakit inline styles
      ,
      style: {
        width: 'auto',
        height: 'auto'
      },
      children: /*#__PURE__*/_jsx(Icon, {
        icon: radioCheck,
        size: 24
      })
    }), /*#__PURE__*/_jsxs(Styled.DropdownMenuItemContentWrapper, {
      children: [/*#__PURE__*/_jsx(Styled.DropdownMenuItemChildrenWrapper, {
        children: children
      }), suffix && /*#__PURE__*/_jsx(Styled.ItemSuffixWrapper, {
        children: suffix
      })]
    })]
  });
});
