// @wordpress/components/build/private-apis.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.privateApis = void 0;
var _utils = require("./popover/utils");
var _menu = require("./menu");
var _contextSystemProvider = require("./context/context-system-provider");
var _theme = _interopRequireDefault(require("./theme"));
var _tabs = require("./tabs");
var _strings = require("./utils/strings");
var _withIgnoreImeEvents = require("./utils/with-ignore-ime-events");
var _lockUnlock = require("./lock-unlock");
var _badge = _interopRequireDefault(require("./badge"));
var _calendar = require("./calendar");
var _validatedFormControls = require("./validated-form-controls");
var _picker = require("./color-picker/picker");
/**
 * Internal dependencies
 */

const privateApis = exports.privateApis = {};
(0, _lockUnlock.lock)(privateApis, {
  __experimentalPopoverLegacyPositionToPlacement: _utils.positionToPlacement,
  ComponentsContext: _contextSystemProvider.ComponentsContext,
  Tabs: _tabs.Tabs,
  Theme: _theme.default,
  Menu: _menu.Menu,
  kebabCase: _strings.kebabCase,
  withIgnoreIMEEvents: _withIgnoreImeEvents.withIgnoreIMEEvents,
  Badge: _badge.default,
  normalizeTextString: _strings.normalizeTextString,
  DateCalendar: _calendar.DateCalendar,
  DateRangeCalendar: _calendar.DateRangeCalendar,
  TZDate: _calendar.TZDate,
  Picker: _picker.Picker,
  ValidatedInputControl: _validatedFormControls.ValidatedInputControl,
  ValidatedCheckboxControl: _validatedFormControls.ValidatedCheckboxControl,
  ValidatedNumberControl: _validatedFormControls.ValidatedNumberControl,
  ValidatedSelectControl: _validatedFormControls.ValidatedSelectControl,
  ValidatedRadioControl: _validatedFormControls.ValidatedRadioControl,
  ValidatedTextControl: _validatedFormControls.ValidatedTextControl,
  ValidatedTextareaControl: _validatedFormControls.ValidatedTextareaControl,
  ValidatedToggleControl: _validatedFormControls.ValidatedToggleControl,
  ValidatedToggleGroupControl: _validatedFormControls.ValidatedToggleGroupControl
});
